src/gui/painting/qprintengine_pdf.cpp

```diff
@@ -833,7 +833,31 @@ class jpg_header_reader {
 
 };
 
+bool QPdfEnginePrivate::smartCompressDeflate(const char *source, int size, QByteArray& compressed, unsigned long maxSize)
+{
+    if (maxSize == 0) {
+        maxSize = ::compressBound(size);
+    }
 
+    const unsigned long initialBufferSize = 10 * 1024 * 1024;
+
+    unsigned long destLen = min(initialBufferSize, maxSize);
+
+    while (true) {
+        compressed.resize(destLen);
+        if (Z_OK == ::compress(reinterpret_cast<Bytef *>(compressed.data()), &destLen, reinterpret_cast<const Bytef *>(source), size)) {
+            compressed.truncate(destLen);
+            break;
+        }
+        if (static_cast<unsigned long>(compressed.size()) >= maxSize) {
```

Comment:
According to [the manual](http://www.zlib.net/manual.html#Utility), `compressBound` returns an upper bound on the compressed size after the `compress` call. So not really sure what this is supposed to protect against? `Z_BUF_ERROR` is obviously not going to happen, so maybe you should check for `Z_MEM_ERROR` directly if that was what you had in mind.